### PR TITLE
Allow duplicate CMS attributes

### LINKED ISSUE #22266: It is no longer possible to insert a second identical unsigned attribute into a CMS/CAdES signature
Hi All,

I'm using OpenSSL 3.0.11 (previously 3.0.8).
Since this change https://github.com/openssl/openssl/pull/21505, it is no longer possible to insert a second identical unsigned attribute into a CMS/CAdES signature (like many archiveTimestamp attributes for example).

The problem lies in the CMS_unsigned_add1_attr_by_NID function, which relies on X509at_add1_attr_by_NID.

Thanks.

### PR DESCRIPTION
**Pull request by t8m**
Fixes regression introduced with https://github.com/openssl/openssl/pull/21505

Fixes #22266

### CODE REVIEW

**Review by slontis — COMMENTED** [review]
This requires tests...
cms_att.c contains a list of cms_attribute_properties and all of these except countersignature have the flags for ONLY_ONE|ONLY_ONE_ATTR_VALUE
These are described in RFC5652 Section 11.
So in most cases duplicate attributes are not allowed.

**Comment by t8m** [comment]
We currently do not have any tests on adding CMS attributes at all. I am not against adding one but this is currently not my priority as I am pretty swamped with other work. However it seems to me blocking this fix on missing tests would be wrong so I set `tests: deferred`.

We will have to review the closed PRs with `tests: deferred` set at some point.

**Comment by slontis** [comment]
@t8m FYI: See the question I asked in https://github.com/openssl/openssl/pull/21505.

**Comment by slontis** [comment]
I could try to add some tests if I have time.. What should happen if you try to add multiple attributes that are not allowed to be duplicated? Does this get deferred until later in the attribute check that happens in sign or verify, or should this happen as the attribute get added? (That was why I was asking about tests).

**Comment by t8m** [comment]
> I could try to add some tests if I have time.. What should happen if you try to add multiple attributes that are not allowed to be duplicated? Does this get deferred until later in the attribute check that happens in sign or verify, or should this happen as the attribute get added? (That was why I was asking about tests).

It is clear that this won't happen when the attribute is added - there is no extra check for that flag given the simplicity of these functions. This is checked when the CMS is signed - only those known attributes are checked, anything else is free-form.

File: crypto/x509/x509_att.c [hunk 1]
```diff
@@ -119,19 +115,68 @@ STACK_OF(X509_ATTRIBUTE) *X509at_add1_attr(STACK_OF(X509_ATTRIBUTE) **x,
     return NULL;
 }
 
+STACK_OF(X509_ATTRIBUTE) *X509at_add1_attr(STACK_OF(X509_ATTRIBUTE) **x,
+                                           X509_ATTRIBUTE *attr)
+{
+    if (x == NULL || attr == NULL) {
```
Comment (slontis) [inline crypto/x509/x509_att.c:121]: Do we care that it does this check twice?

Comment (t8m) [reply crypto/x509/x509_att.c:121]: IMO no.

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master, 3.2, 3.1, and 3.0 branches. Thank you for the reviews.

**Pull request closed without merging**